File: 0907-koko-eating-bananas/0907-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)

        result = r

        while l <= r:
            mid = (l + r) // 2
            count = 0

            for pile in piles:
                count += math.ceil(pile / mid)
            
            if count <= h:
                result = mid
                r = mid - 1
            else:
                l = mid + 1
        
        return result

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.cpp
class Solution {
public:
    bool possible(int speed, const vector<int>& piles, int h) {
        int hour = 0;
        for (int pile : piles) {
            hour += ceil(pile / (double)speed);
        }
        return hour <= h;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 1;
        int r = *max_element(piles.begin(), piles.end());

        while (l < r) {
            int mid = (l + r) / 2;
            if (possible(mid, piles, h)) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }

        return l;
    }
};